### ISSUE TITLE
CryptoProviderFactory.Default has a reference to a disposable object

### ISSUE THREAD

**Issue by imokaythanks** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#1650]
When `CryptoProviderFactory` creates an `InMemoryCryptoProviderCache`
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/b45a5b4cf4e918ca2b2c12ff1951f1be7b592bac/src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs#L110
Two `EventBasedLRUCache` are created
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/b45a5b4cf4e918ca2b2c12ff1951f1be7b592bac/src/Microsoft.IdentityModel.Tokens/InMemoryCryptoProviderCache.cs#L92-L93
Each of these fire off a `OnStart` task.
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/b45a5b4cf4e918ca2b2c12ff1951f1be7b592bac/src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs#L81

We have a requirement that unit tests cannot leave tasks hanging around, because if they throw an unobserved exception, it can cause an unrelated test to fail. So a test that touches `CryptoProviderFactory.Default` can fail this requirement if it finishes before the `OnStart` tasks.

A solution is to call
```
var cache = (InMemoryCryptoProviderCache)CryptoProviderFactory.Default.CryptoProviderCache;
cache.Dispose();
```
before exiting the test.

But this requires some internal knowledge of the `CryptoProviderCache` type.  Since `InMemoryCryptoProviderCache` implements `IDisposable`, and `CryptoProviderFactory` creates it, I think it should dispose it or implement `IDisposable` itself.

**Comment by jmprieur** [comment]
@brentschmaltz @GeoK @mafurman : how do we triage this one? Which label should it have?

**Comment by mafurman** [comment]
The CryptoProviderFactory currently doesn't implement IDisposable since the type of the cache property on it is the CryptoProviderCache abstract class, which doesn't implement IDisposable.

We should probably either add IDisposable to the the abstract class or cast the object to an instance of InMemoryCryptoProviderCache. Either way, the CryptoProviderFactory should implement IDisposable itself.

@jmprieur I've labeled this as 'customer reported' and 'bug' for now.

**Comment by brentschmaltz** [comment]
@jmprieur @mafurman the tension here is that the user usually never creates the CryptoProviderFactory and it seems odd to have Dispose on a Factory. We may want to think about using finalizers.

**Comment by dannybtsai** [comment]
While it seems odd to dispose a factory object, the finalizer might not work as the program has no control over when it will be called.
It seems in order to completely address the issue @imokaythanks mentioned in unit tests, we need to add:
- the IDisposable interface to the CryptoProviderFactory so callers, using either the CryptoProviderFactory.Default or new CryptoProviderFactory(), can dispose all objects including the tasks at the end of tests. The danger is that the CryptoProviderFactory.Default can be disposed by accident.
- the finalizer to ensure all objects and tasks will be disposed eventually;
- Lazy initialization of the CryptoProviderFactory.Default instance to provide the option of not creating the instance at all for tests while supporting the existing code using this instance.
- 
@imokaythanks, can you share how you verify any remaining running tasks at the end of your tests?
And in your post, you can work-around the issue by replacing

var cache = (InMemoryCryptoProviderCache)CryptoProviderFactory.Default.CryptoProviderCache;
cache.Dispose();

with

using(var cache =new InMemoryCryptoProviderCache(new CryptoProviderCacheOptions(), TaskCreationOptions.None))
{
  var factory = new CryptoProviderFactory(cache);
  ...
}

if that makes sense, at least for now.

**Comment by brentschmaltz** [comment]
@imokaythanks (love the name) we are thinking we will stop the task when the cache is empty.
@dannybtsai we may want to think about only turning on the task when the cache gets to a certain size.

**Comment by imokaythanks** [comment]
@dannybtsai We call Task.GetActiveTasks with reflection after each test (lol). I do think it's a reasonable unit test requirement though.
https://referencesource.microsoft.com/mscorlib/a.html#fa9c7653b4c2d13c

Thanks all.

**Comment by dannybtsai** [comment]
> @dannybtsai We call Task.GetActiveTasks with reflection after each test (lol). I do think it's a reasonable unit test requirement though.
> https://referencesource.microsoft.com/mscorlib/a.html#fa9c7653b4c2d13c
> 
> Thanks all.

@imokaythanks, I tried the Task.GetActiveTasks() with reflection as you suggested and it has been working for the last few days but now started to fail in VS:
            MethodInfo methodInfo = typeof(Task).GetTypeInfo().GetMethod("GetActiveTasks", BindingFlags.NonPublic | BindingFlags.Static);
            var activeTasksAtStart = methodInfo.Invoke(this, new object[] { }) as Task[];


Both typeof(Task).GetTypeInfo().GetMethod() and typeof(Task).GetMethod() are returning null now.
Have you seen the problem before or how do you do it in your tests?

**Comment by imokaythanks** [comment]
@dannybtsai  That is how we call it. Not sure how it can return null, it appears to be an empty list of there are no tasks. I did get the failure a few times but I could see how this is a race condition and sometimes would not fail.

I didn't intend for this issue to get caught up in the details of our test. I raised it because it seemed to expose an improper use of  disposable objects that maybe could cause other issues.

**Comment by dannybtsai** [comment]
@imokaythanks, a new version, 6.12.0, has been released to address this issue (see https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/Caching-in-Microsoft.IdentityModel). Please take a look and let me know how it works.
BTW, calling the Task.GetActiveTasks() using reflection seems to work in some environments but failed in others and it is not clear why, maybe it has something to do with the run-time versions.

**Comment by mackie1001** [comment]
Note that the fix in 6.12.0 also appears to fix this issue: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1630
